### PR TITLE
Network configuration for Docker services and usage of existing networks

### PR DESCRIPTION
**Pull request by alexfdezsauco**
#282 Add configuration option to enable the containers usage of an existing network

### CODE REVIEW

File: src/Microsoft.Tye.Hosting/DockerRunner.cs [hunk 1]
```diff
@@ -51,8 +51,32 @@ public async Task StartAsync(Application application)
 
             string? dockerNetwork = null;
 
+            if (!string.IsNullOrWhiteSpace(application.Network))
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/DockerRunner.cs:54]: IsNullOrWhiteSpace -> IsNullOrEmpty

File: src/Microsoft.Tye.Hosting/DockerRunner.cs [hunk 2]
```diff
@@ -94,7 +118,7 @@ public async Task StopAsync(Application application)
 
             await Task.WhenAll(tasks);
 
-            if (application.Items.TryGetValue("dockerNetwork", out var dockerNetwork))
+            if (string.IsNullOrWhiteSpace(application.Network) && application.Items.TryGetValue("dockerNetwork", out var dockerNetwork))
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/DockerRunner.cs:121]: Same

**Comment by davidfowl** [comment]
This change also needs a test.

File: src/Microsoft.Tye.Hosting/DockerRunner.cs (hunk 1 above)
Comment (alexfdezsauco) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs]: Fixed

File: src/Microsoft.Tye.Hosting/DockerRunner.cs (hunk 2 above)
Comment (alexfdezsauco) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs]: Fixed

File: src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs [hunk 3]
```diff
@@ -21,6 +21,8 @@ public class ConfigApplication
 
         public string? Registry { get; set; }
 
+        public string? Network { get; set; }
```
Comment (davidfowl) [inline src/Microsoft.Tye.Core/ConfigModel/ConfigApplication.cs:24]: @rynowak how do you feel about this? I don't love it being so top level and being specific to containers but I don't have an alternative yet.

**Comment by alexfdezsauco** [comment]
> Can you also write a test?

Yes, I can. I'm getting related with Tye testing approach (kind of integration tests). 

Basically I have to test if a network exists the docker execution doesn't create a new one and doesn't dispose it at the end. I'm thinking to create a docker only yaml file but it will require at least one image (or to download such image) to run.

**Comment by davidfowl** [comment]
> Basically I have to test if a network exists the docker execution doesn't create a new one and doesn't dispose it at the end. I'm thinking to create a docker only yaml file but it will require at least one image (or to download such image) to run.

Model it after the volume test https://github.com/dotnet/tye/blob/6671da72c0f7ef83ddabcd8d44e2e0ebb7d0e3f2/test/E2ETest/TyeRunTests.cs#L193.

Create a network at the beginning of the test with a random name and programmatically change the model to use it. Then after it works, make a rest API call to get the container id and then use docker inspect to make sure the container is on the right network (or maybe use the REST API that tye exposes).

File: test/E2ETest/E2ETest.csproj [hunk 4]
```diff
@@ -40,6 +40,10 @@
     <None Remove="testassets\generate\single-project.yaml" />
     <None Remove="testassets\init\frontend-backend.yaml" />
     <None Remove="testassets\init\multi-project.yaml" />
+    <None Remove="testassets\projects\network-test\appsettings.Development.json" />
```
Comment (davidfowl) [inline test/E2ETest/E2ETest.csproj:43]: Remove?

File: test/E2ETest/testassets/projects/network-test/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            if (env.IsDevelopment())
+            {
+                app.UseDeveloperExceptionPage();
+            }
+
+            app.UseRouting();
+
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapGet("/", async context =>
```
Comment (davidfowl) [inline test/E2ETest/testassets/projects/network-test/Startup.cs:34]: Change the sample so it isn't using volumes. That isn't what's being tested here.

File: test/E2ETest/TyeRunTests.cs [hunk 6]
```diff
@@ -240,6 +244,80 @@ public async Task DockerNamedVolumeTest()
             await ProcessUtil.RunAsync("docker", $"volume rm {volumeName}");
         }
 
+        [ConditionalFact]
+        [SkipIfDockerNotRunning]
+        public async Task DockerNetworkAssignmentTest()
+        {
+            using var projectDirectory = CopyTestProjectDirectory("network-test");
```
Comment (davidfowl) [inline test/E2ETest/TyeRunTests.cs:251]: I would actually use the frontend-backend example. That does container to container communication and is written for you already. Forgot to mention that earlier.

File: src/Microsoft.Tye.Hosting/DockerRunner.cs [hunk 7]
```diff
@@ -51,8 +51,32 @@ public async Task StartAsync(Application application)
...
+
+                if (!string.IsNullOrWhiteSpace(dockerNetworkResult.StandardOutput))
+                {
+                    _logger.LogInformation("The specified network {Network} exists", application.Network);
+
+                    dockerNetwork = application.Network;
+                }
+                else
+                {
+                    _logger.LogInformation("The specified network {Network} doesn't exist.", application.Network);
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/DockerRunner.cs:72]: LogError?

File: test/E2ETest/TyeRunTests.cs [hunk 8]
```diff
@@ -21,6 +21,10 @@
 
 namespace E2ETest
 {
+    using Newtonsoft.Json;
+
+    using JsonSerializer = System.Text.Json.JsonSerializer;
```
Comment (davidfowl) [inline test/E2ETest/TyeRunTests.cs:26]: Undo?

File: test/E2ETest/TyeRunTests.cs (hunk 6 above)
Comment (davidfowl) [reply test/E2ETest/TyeRunTests.cs:251]: Let me be more specific, there's a test called `FrontendBackendRunTestWithDocker`. Copy that test and add the custom code to create a network.

**Comment by alexfdezsauco** [comment]
Fix according with the revision. But there is an issue with whitespace formatting, but not sure why.

**Comment by alexfdezsauco** [comment]
Fixed

**Comment by davidfowl** [comment]
Did you remove the check to make sure the app is working? Keep the original assertions...

**Comment by alexfdezsauco** [comment]
Fixed

**Comment by davidfowl** [comment]
I’d recommend pinging the backend before the frontend to reduce potential flakiness. This was an intentional change that was made in previous PRs because of random failures

**Comment by davidfowl** [comment]
Thanks!

**Pull request merged**